pkg/repository/git.go

```diff
@@ -122,53 +120,60 @@ func (r gitRepoController) getSHA() (string, error) {
 	return head.Hash().String(), nil
 }
 
-func (r gitRepoController) getTreeSHA(context string) (string, error) {
-	repo, err := r.repoGetter.get(r.path)
-	if err != nil {
-		return "", fmt.Errorf("failed to get repository: %w", err)
-	}
+type commitResponse struct {
+	commit string
+	err    error
+}
 
-	ref, err := repo.Head()
-	if err != nil {
-		return "", fmt.Errorf("failed to get HEAD from repo: %w", err)
-	}
+func (r gitRepoController) GeLatestDirCommit(contextDir string) (string, error) {
+	// We use context.TODO() in a few places to call isClean, so let's make sure
+	// we set proper internal timeouts to not leak goroutines
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
 
-	commit, err := repo.CommitObject(ref.Hash())
-	if err != nil {
-		return "", fmt.Errorf("failed to get commit object from reference: %w", err)
-	}
-	tree, err := commit.Tree()
-	if err != nil {
-		return "", fmt.Errorf("failed to get tree from commit: %w", err)
-	}
+	timeoutCh := make(chan struct{})
+	ch := make(chan commitResponse)
 
-	if context == "." {
-		return tree.Hash.String(), nil
-	}
+	timeoutErr := errors.New("timeout exceeded")
```

Comment:
considering we have `timeoutErr` in two funcs in this file, should we move it at the top level of thisfile?